napari/layers/graph/graph.py

```diff
@@ -0,0 +1,339 @@
+
+    def _get_ndim(self) -> int:
+        """Determine number of dimensions of the layer."""
+        return self.data.ndim
+
+    def _make_slice_request_internal(
+        self, slice_input: _SliceInput, dims_indices: ArrayLike
+    ) -> _GraphSliceRequest:
+        return _GraphSliceRequest(
+            dims=slice_input,
+            data=self.data,
+            dims_indices=dims_indices,
+            out_of_slice_display=self.out_of_slice_display,
+            size=self.size,
+        )
+
+    def _update_slice_response(self, response: _GraphSliceResponse) -> None:  # type: ignore[override]
+        super()._update_slice_response(response)
+        self._edges_indices_view = response.edges_indices
+
+    @property
+    def _view_edges_coordinates(self) -> np.ndarray:
+        return self.data._coords[self._edges_indices_view][
+            ..., self._slice_input.displayed
+        ]
+
+    def add(
+        self, coords: ArrayLike, indices: Optional[ArrayLike] = None
+    ) -> None:
+        """Adds nodes at coordinates.
+        Parameters
+        ----------
+        coords : sequence of indices to add point at
+        indices : optional indices of the newly inserted nodes.
+        """
+        coords = np.atleast_2d(coords)
+        if indices is None:
+            new_starting_idx = self.data._buffer2world.max() + 1
+            indices = np.arange(
+                new_starting_idx, new_starting_idx + len(coords)
+            )
+
+        indices = np.atleast_1d(indices)
+
+        if len(coords) != len(indices):
+            raise ValueError(
+                trans._(
+                    'coordinates and indices must have the same length. Found {coords_size} and {idx_size}',
+                    coords_size=len(coords),
+                    idx_size=len(indices),
+                )
+            )
+
+        prev_size = self.data.n_allocated_nodes
+
+        for idx, coord in zip(indices, coords):
+            self.data.add_nodes(idx, coord)
+
+        self._data_changed(prev_size)
+
+    def remove_selected(self):
+        """Removes selected points if any."""
+        if len(self.selected_data):
+            indices = self.data._buffer2world[list(self.selected_data)]
+            self.remove(indices)
+            self.selected_data = set()
+
+    def remove(self, indices: ArrayLike) -> None:
+        """Removes nodes given their indices."""
+        indices = np.atleast_1d(indices)
+        if indices.ndim > 1:
+            raise ValueError(
+                trans._(
+                    "Indices for removal must be 1-dim. Found {ndim}",
+                    ndim=indices.ndim,
+                )
+            )
+
+        prev_size = self.data.n_allocated_nodes
+        # descending order
+        indices = np.flip(np.sort(indices))
+
+        # it got error missing __iter__ attribute, but we guarantee by np.atleast_1d call
+        for idx in indices:  # type: ignore[union-attr]
+            self.data.remove_node(idx)
```

Comment:
No, we can implement it, but we will have to duplicate some existing code by moving this method the `BaseGraph` to the directed and undirected classes because it calls `_remove_incident_edges`, which is overridden by their children, [reference](https://github.com/napari/napari-graph/blob/4c47282dd4d9959be6fb77bb706158e8448d31f8/src/napari_graph/base_graph.py#L420)

We can always convert it to a non-jitted `remove_nodes`  method.

Reply:
I think a non-jitted version would come a long way with the `buffer_indices` parameter, we could avoid the private attribute access mentioned in the comment above and we could simplify the graph layer `remove` method implementation. Do you agree?